tmt/steps/provision/__init__.py

```diff
@@ -1788,6 +1789,84 @@ def fetch_logs(
         for log in guest_logs:
             log.store(logger, dirpath, log.name)
 
+    def _construct_mkdtemp_command(
+        self,
+        prefix: Optional[str] = None,
+        template: Optional[str] = None,
+        parent: Optional[Path] = None,
+    ) -> Command:
+        template = template or 'tmp.XXXXXXXXXX'
+
+        if prefix is not None:
+            template = f'{prefix}{template}'
+
+        options: list[str] = ['--directory']
+
+        if parent is not None:
+            options += ['-p', str(parent)]
+
+        return Command(*('mktemp', *options, template))
+
+    @contextlib.contextmanager
+    def mkdtemp(
+        self,
+        # Suffix is not supported everywhere, namely Alpine does not
+        # recognize it, and even requires template to end with `XXXXXX`.
+        # Therefore not supporting this option - in the future, someone
+        # may need it, fix it for all distros, and uncomment the
+        # parameter.
+        # suffix: Optional[str] = None,
+        prefix: Optional[str] = None,
+        template: Optional[str] = None,
+        parent: Optional[Path] = None,
+    ) -> Iterator[Path]:
+        """
+        Create a temporary directory.
+
+        Modeled after :py:func:`tempfile.mkdtemp`, but creates the
+        temporary directory on the guest, by invoking ``mktemp -d``. The
+        implementation may slightly differ, but the temporary directory
+        should be created safely, without conflicts, and it should be
+        accessible only to user who created it.
+
+        Since the caller is responsible for removing the directory, it
+        is recommended to use it as a context manager, just as one would
+        use :py:func:`tempfile.mkdtemp`; leaving the context will remove
+        the directory:
+
+        .. code-block:: python
+
+            with guest.mkdtemp() as path:
+                ...
+
+        :param prefix: if set, the directory name will begin with this
+            string.
+        :param template: if set, the directory name will follow the
+            given naming scheme: the template must end with 6
+            consecutive ``X``s, i.e. ``XXXXXX``. All ``X`` letters will
+            be replaced with random characters.
+        :param parent: if set, new directory will be created under this
+            path. Otherwise, the default directory is used.
+        """
+
+        output = self.execute(
+            self._construct_mkdtemp_command(prefix=prefix, template=template, parent=parent)
+        )
+
+        if not output.stdout:
+            raise GeneralError(f"Failed to create temporary directory on guest: {output.stderr}")
+
+        path = Path(output.stdout.strip())
+
+        try:
+            yield path
+
+        except Exception as exc:
```

Comment:
What do you think about keeping the directory on failure? Not sure if it should be optional, and if yes, opt-in or opt-out.
```suggestion
        except Exception as exc:
            if not keep_on_error:
                self.execute(Command('rm', '-rf', path))
            else:
                self.debug(f"Keeping temporary directory {path} due to error")
```

Reply:
my IDE gave me this unit test to cover it fwiw:
```
diff --git a/tests/unit/test_guest.py b/tests/unit/test_guest.py
index 5058982b..0c981901 100644
--- a/tests/unit/test_guest.py
+++ b/tests/unit/test_guest.py
@@ -100,6 +100,7 @@ def test_mkdtemp(
 ) -> None:
     guest.execute(ShellScript('mkdir -p /tmp/qux'))
 
+    # Test basic functionality
     with guest.mkdtemp(
         prefix='bar',
         template='XXXXXXbazXXXXXX',
@@ -107,7 +108,43 @@ def test_mkdtemp(
     ) as path:
         guest.execute(ShellScript(f'ls -al {path}'))
 
+    # Verify directory was removed
     with pytest.raises(RunError) as exc_context:
         guest.execute(ShellScript(f'ls -al {path}'))
 
     assert re.match(r'(?i).*?No such file or directory.*', exc_context.value.stderr or '')
+
+
+@pytest.mark.containers
+@pytest.mark.parametrize(
+    ('container',),  # noqa: PT006
+    [(TEST_CONTAINERS['CONTAINER_FEDORA_41'],)],  # Test only on one container to save time
+    indirect=["container"],
+    ids=[TEST_CONTAINERS['CONTAINER_FEDORA_41'].url],
+)
+def test_mkdtemp_keep_on_error(
+    container: ContainerData,
+    guest: GuestContainer,
+    root_logger: Logger,
+    caplog: _pytest.logging.LogCaptureFixture,
+) -> None:
+    error_message = "Simulated error for testing"
+    
+    try:
+        with guest.mkdtemp(
+            prefix='keep-me-',
+            keep_on_error=True,
+        ) as path:
+            # Record the path for later verification
+            temp_path = path
+            # Create a marker file to verify the directory exists
+            guest.execute(ShellScript(f'touch {path}/marker'))
+            # Simulate an error
+            raise RuntimeError(error_message)
+    except RuntimeError as exc:
+        assert str(exc) == error_message
+        # Verify the directory still exists with our marker
+        output = guest.execute(ShellScript(f'ls -la {temp_path}/marker'))
+        assert output.stdout.strip() != ""
+        # Clean up manually for the test
+        guest.execute(Command('rm', '-rf', temp_path))
```

Reply:
@martinhoyer ping